Add deployment-target --print flag for Apple targets

BlackHoleFox opened the pull request:
This is very useful for crates that need to know what the Apple OS deployment target is for their build scripts or inside of a build environment. Right now, the defaults just get copy/pasted around the ecosystem since they've been stable for so long. But with #104385 in progress, that won't be true anymore and everything will need to move. Ideally whenever it happens again, this could be less painful as everything can ask the compiler what its default is instead.

To show examples of the copy/paste proliferation, here's some crates and/or apps that do:
- [cc](https://github.com/rust-lang/cc-rs/pull/708/files), Soon
-  [mac-notification-sys](https://github.com/h4llow3En/mac-notification-sys/pull/46/files#diff-d0d98998092552a1d3259338c2c71e118a5b8343dd4703c0c7f552ada7f9cb42R10-R12)
- [PyO3](https://github.com/PyO3/maturin/blob/ccb02d1aa1cc41e82a3572a3c8b35cace15f3e78/src/target.rs#L755-L758)
- [Anki](https://github.com/ankitects/anki/blob/613b5c1034cc9943f3f68d818ae22b2e0acec877/build/runner/src/bundle/artifacts.rs#L49-L54)
- [jsc-rs](https://github.com/Brooooooklyn/jsc-rs/blob/37767267568fb2de62fc441473e7d158dd980520/xtask/src/build.rs#L402-L405)
... and probably more that a simple GitHub codesearch didn't see

BlackHoleFox committed:
Add deployment-target --print flag for Apple targets

BlackHoleFox commented:
@rustbot label O-macos O-ios

BlackHoleFox commented:
So far I'm unsure of the flag name. `minimum-version` might be better but `deployment-version` might be easier to find?

compiler-errors commented:
r? compiler

oli-obk commented:
r? @oli-obk

oli-obk commented on compiler/rustc_target/src/spec/apple_base.rs:
```diff
@@ -173,7 +173,23 @@ pub fn opts(os: &'static str, arch: Arch) -> TargetOptions {
+    format!("{major}.{minor}")
+}
+
+fn from_set_deployment_target(var_name: &str) -> Option<(u32, u32)> {
     let deployment_target = env::var(var_name).ok();
     deployment_target
         .as_ref()
```
preexisting, but this function deserves a few question mark operators and named bindings to become more readable.

BlackHoleFox replied:
It should be better now. Primarily, things are named now to make the parsing more reasonable to read over.

oli-obk commented:
This adds a new insta-stable compiler flag. This seems like a good solution to the problem for me.

@rfcbot fcp merge

lcnr commented on compiler/rustc_driver/src/lib.rs:
```diff
@@ -743,6 +743,18 @@ fn print_crate_info(
+                } else {
+                    early_error(
+                        ErrorOutputType::default(),
+                        "Only Apple targets currently support deployment version info",
```
```suggestion
                        "only Apple targets currently support deployment version info",
```

can you add a `ui` test for this error is emitted

BlackHoleFox replied:
Yeah, that works. There's one now that uses the x64 GNU Linux target to test the error branch. The casing is fixed now as well.

petrochenkov commented on src/doc/rustc/src/command-line-arguments.md:
```diff
@@ -248,8 +248,14 @@ The valid types of print values are:
+- `deployment-target` - The currently selected [deployment target] (or minimum OS version)
+  for the selected Apple platform target. This value can be used or passed along to other
+  components alongside a Rust build that need this information, such as C compilers.
+  This returns rustc's minimum supported deployment target if no `_DEPLOYMENT_TARGET` variable
```
```suggestion
  This returns rustc's minimum supported deployment target if no `*_DEPLOYMENT_TARGET` variable
```

petrochenkov commented on compiler/rustc_target/src/spec/apple_base.rs:
```diff
@@ -173,7 +173,23 @@ pub fn opts(os: &'static str, arch: Arch) -> TargetOptions {
+        "ios" => ios_deployment_target(),
+        "watchos" => watchos_deployment_target(),
+        "tvos" => tvos_deployment_target(),
+        _ => unreachable!("unknown Apple target OS"),
```
This arm is reachable with custom target specs.
I suggest returning an `Option` here, and relying on it in the driver instead of using `sess.target.is_like_osx`.

BlackHoleFox replied:
Ah, thanks, and good to know. Moved up to `rustc_driver`. For the future, why is doing it there better? Do custom target specs go through a different codepath that wouldn't panic?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3747027693/jobs/6362871210) [(plain)](https://github.com/rust-lang/rust/commit/95075d942aa35039effcb36a4d29cff0e4067c04/checks/10224733264/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Finished release [optimized] target(s) in 13.69s
tidy check
tidy: Skipping binary file check, read-only filesystem
Checking which error codes lack tests...
tidy error: following path contains more than 939 entries, you should move the test to some relevant subdirectory (current: 940): /checkout/src/test/ui
* highest error code: E0791
* highest error code: E0791
/checkout/src/test/ui/deployment-target/invalid-target.rs: revision [unspecified] should specify `needs-llvm-components:` as it has `--target` set
Found 0 error(s) in error codes
Done!
* 392 features
some tidy checks failed
```

</details>

BlackHoleFox commented:
There's now also a test for valid use of the new flag. It isn't very picky about the output beyond requiring that it only contains `{Digit}.0`. Imo its not worth hardcoding rustc's minimum macOS version in another place because there's already compile tests that verify it. That and it makes updating it later more tedious.

bors commented:
:umbrella: The latest upstream changes (presumably #106087) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

thomcc commented:
This has been discussed many times, <https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Pre-RFC.3A.20CFG.20OS.20Version> is one of the more recent efforts. Anecdotally, they seem to tend to hit snags as they often come in with only planned support for a single OS, and wish to leave the others as future work, which may not be viable (in the case of Apple targets, given that we already have support for this in the compiler as it's not exactly an optional aspect of those targets, my feeling is that it seems like it would need to be addressed).

Regardless, I'm not sure these two overlap that much -- the cfgs that have been suggested are in the form of an `os_version_min` predicate, and it isn't clear how that will come through to build scripts (FWIW, I'd love to be able to use this in `cc`).

BlackHoleFox commented:
Like what @thomcc said, I think this should exist in parallel with a hypothetical `os_version_min`. One's for your own Rust code and the other for anything (`clang`, `gcc`, `swiftc`, etc) running inside of a build script.

> and it isn't clear how that will come through to build scripts 

In the same grouping as the other `cfg!(platform_stuff)` conditionals, I don't think it would.

nagisa commented:
> Do you mean something like deployment_target=10.x or similar? [...] Do you imagine something simple like above or would it be much more complex to the extent you need a crate like rustc_version to do anything?

Something as simple as `deployment_target=10.x` is kind of exactly what I’m thinking of, again drawing parallels with the output of `--print=cfg`.

madsmtm commented:
I'm not sure I understand why `cargo` doesn't just set the `MACOSX_DEPLOYMENT_TARGET` environment variable (and similar) when executing build scripts?

That way, code that already checks the standard environment variables wouldn't need to be updated, and everything would just work?

oli-obk commented:
@rustbot author

needs a rebase.

BlackHoleFox commented:
Oh wow, this one fell out of my head. I'll rebase this and finish addressing @nagisa's comment to update the output format.

> I'm not sure I understand why cargo doesn't just set the MACOSX_DEPLOYMENT_TARGET environment variable (and similar) when executing build scripts?

@madsmtm I was kind of thinking this would be the start of better deployment target support/visibility in Rust tooling. Just having Cargo set it for build scripts wouldn't be enough afaik since `rustc` needs it as well as whatever `cc` you're using in some `build.rs`.

The main reason is that I wanted a way to keep the Rust Apple ecosystem easily in sync with the minimum supported OSes. Since `rustc` defines these, and not Cargo, adding this flag to expose either the default _or_ whatever `$PLATFORM_DEPLOYMENT_TARGET` was set to seemed to be a good start. Today nothing exists for that so the 10.12 `rustc` bump I'm working on is going to throw a lot of projects off.

So after this flag theoretically lands, we'd have a way to go from `env var/compiler default` --> Cargo. Cargo could (and probably should once some other RFCs land) take into account its own configuration file's minimum platform version as well (akin to XCode's config files supporting this too). With either of those version sources, Cargo could then inject a deterministic environment variable into build scripts to give user-controlled compilers the needed info. While a smaller point, some other build systems don't support Cargo at all either, so again I think its best this info can always be sourced from `rustc` regardless of how Cargo ends up exposing it to `build.rs`.

BlackHoleFox commented:
Rebased and updated to apply @nagisa's suggestion of making the print output self-describing, like the `cfg` ones:
```
rustc +stage1  --print deployment-target
deployment_target=11.0

rustc +stage1 --target x86_64-apple-darwin --print deployment-target
deployment_target=10.7
```

@rustbot ready

bors commented:
:umbrella: The latest upstream changes (presumably #111169) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 1e45f0885d766c3d5b1279cd9fb9e78f41cd4ab2 has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 1e45f0885d766c3d5b1279cd9fb9e78f41cd4ab2 -->
<!-- homu: {"type":"Approved","sha":"1e45f0885d766c3d5b1279cd9fb9e78f41cd4ab2","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

JohnTitor commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/111219#issuecomment-1535589391
@bors r-

BlackHoleFox commented:
Should be ready again, the UI test wasn't flexible enough before to handle different deployment versions. I fixed that and tried it out locally with multiple `MACOSX_DEPLOYMENT_VERSIONS`.

@rustbot ready

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit a427d418fd0004f12a7d59899e6994238daed9aa has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk a427d418fd0004f12a7d59899e6994238daed9aa -->
<!-- homu: {"type":"Approved","sha":"a427d418fd0004f12a7d59899e6994238daed9aa","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
